lib/Core/SQLiteBuildDB.cpp

```diff
@@ -614,17 +616,17 @@ class SQLiteBuildDB : public BuildDB {
     // FIXME: We could save some reallocation by having a templated SmallVector
     // size here.
     basic::BinaryEncoder encoder{};
-    for (auto keyIDAndFlag: ruleResult.dependencies) {
+    for (auto dependency: ruleResult.dependencies) {
       // Map the enging keyID to a database key ID
       //
       // FIXME: This is naively mapping all keys with no caching at this point,
       // thus likely to perform poorly.  Should refactor this into a bulk
       // query or a DB layer cache.
-      auto dbKeyID = getKeyID(keyIDAndFlag.keyID, error_out);
+      auto dbKeyID = getKeyID(dependency.keyID, error_out);
       if (!error_out->empty()) {
         return false;
       }
-      encoder.write((dbKeyID.value << 1) + keyIDAndFlag.flag);
+      encoder.write((dbKeyID.value << 2) + (dependency.singleUse << 1) + dependency.orderOnly);
```

Comment:
Should this really be writing single use dependencies into the DB?   Wouldn't it be more appropriate to return an error here?   Single use entries by definition shouldn't ever be written to the db.